swiftnav/src/time/gnss.rs

```diff
@@ -0,0 +1,816 @@
+
+        while self.tow >= consts::WEEK_SECS as f64 {
+            self.tow -= consts::WEEK_SECS as f64;
+            self.wn += 1;
+        }
+    }
+
+    /// Adds a duration to the time
+    pub fn add_duration(&mut self, duration: &Duration) {
+        self.tow += duration.as_secs_f64();
+        self.normalize();
+    }
+
+    /// Subtracts a duration from the time
+    pub fn subtract_duration(&mut self, duration: &Duration) {
+        self.tow -= duration.as_secs_f64();
+        self.normalize();
+    }
+
+    /// Gets the difference between this and another time value in seconds
+    pub fn diff(&self, other: &Self) -> f64 {
+        let dt = self.tow - other.tow;
+        dt + (self.wn - other.wn) as f64 * consts::WEEK_SECS as f64
+    }
+
+    /// Convert a [`GpsTime`] into a [`UtcTime`] using the [`UtcParams`] if
+    /// available, or hardcoded leap seconds otherwise
+    fn internal_to_utc(self, params: Option<&UtcParams>) -> UtcTime {
+        // Is it during a (positive) leap second event
+        // Get the UTC offset at the time we're converting
+        let (is_lse, dt_utc) = params.map_or_else(
+            || {
+                (
+                    self.is_leap_second_event_hardcoded(),
+                    self.gps_utc_offset_hardcoded(),
+                )
+            },
+            |p| (self.is_leap_second_event(p), self.gps_utc_offset(p)),
+        );
+
+        let mut tow_utc = self.tow - dt_utc;
+
+        if is_lse {
+            /* positive leap second event ongoing, so we are at 23:59:60.xxxx
+             * subtract one second from time for now to make the conversion
+             * into yyyy/mm/dd HH:MM:SS.sssss format, and add it back later */
+            tow_utc -= 1.0;
+        }
+
+        let mut t_u = GpsTime {
+            wn: self.wn,
+            tow: tow_utc,
+        };
+        t_u.normalize();
+
+        /* break the time into components */
+        let mut u = UtcTime::from_gps_no_leap(t_u);
+
+        if is_lse {
+            assert!(u.hour() == 23);
+            assert!(u.minute() == 59);
+            assert!(u.seconds_int() == 59);
+            /* add the extra second back in*/
+            u.add_second();
+        }
+
+        u
+    }
+
+    /// Converts the GPS time into UTC time
+    ///
+    /// # Panics
+    ///
+    /// This function will panic if the GPS time is not valid
+    pub fn to_utc(self, utc_params: &UtcParams) -> UtcTime {
+        self.internal_to_utc(Some(utc_params))
+    }
+
+    /// Converts the GPS time into UTC time using the hardcoded list of leap
+    /// seconds.
+    ///
+    /// # ⚠️  🦘  ⏱  ⚠️  - Leap Seconds
+    ///
+    /// The hard coded list of leap seconds will get out of date, it is
+    /// preferable to use [`GpsTime::to_utc()`] with the newest set of UTC parameters
+    ///
+    /// # Panics
+    ///
+    /// This function will panic if the GPS time is not valid
+    pub fn to_utc_hardcoded(self) -> UtcTime {
+        self.internal_to_utc(None)
+    }
```

Comment:
Rather than using the `_hardcoded` convention, usually what i see is other crates will call, for example, this function `to_utc` and the one above `to_utc_with_params` or `to_utc_with_config`. Do you feel strongly about this one way or another?